lib/portage/util/file_copy.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+import errno
+import fcntl
+import logging
+import os
+import platform
+import shutil
+import sys
+
+
+logger = logging.getLogger(__name__)
+FICLONE = 0x40049409
+
+
+def _get_chunks(src):
+    if not hasattr(os, "SEEK_DATA"):
+        offset_end = os.lseek(src, 0, os.SEEK_END)
+        yield (0, offset_end)
+        return
+
+    offset_hole = 0
+    while True:
+        try:
+            # Find the next bit of data
+            offset_data = os.lseek(src, offset_hole, os.SEEK_DATA)
+        except OSError as e:
+            offset_end = os.lseek(src, 0, os.SEEK_END)
+
+            if e.errno == errno.ENXIO:
+                if offset_end > offset_hole:
+                    # Hole at end of file
+                    yield (offset_end, 0)
+            else:
+                # SEEK_DATA failed, return the whole file
+                yield (0, offset_end)
+
+            break
+        else:
+            offset_hole = os.lseek(src, offset_data, os.SEEK_HOLE)
+            yield (offset_data, offset_hole - offset_data)
+
+
+def _do_copy_file_range(src, dst, offset, count):
+    while count > 0:
+        # count must fit in ssize_t
+        c = min(count, sys.maxsize)
+        written = os.copy_file_range(src, dst, c, offset, offset)
+        if written == 0:
+            # https://bugs.gentoo.org/828844
+            raise OSError(errno.EOPNOTSUPP, os.strerror(errno.EOPNOTSUPP))
+        offset += written
+        count -= written
+
+
+def _do_sendfile(src, dst, offset, count):
+    os.lseek(dst, offset, os.SEEK_SET)
+    while count > 0:
+        # count must fit in ssize_t
+        c = min(count, sys.maxsize)
+        written = os.sendfile(dst, src, offset, c)
+        offset += written
+        count -= written
+
+
+def _fastcopy(src, dst):
+
+    with (
+        open(src, "rb", buffering=0) as srcf,
+        open(dst, "wb", buffering=0) as dstf,
+    ):
+        srcfd = srcf.fileno()
+        dstfd = dstf.fileno()
+
+        if platform.system() == "Linux":
+            try:
+                fcntl.ioctl(dstfd, FICLONE, srcfd)
+                return
+            except OSError:
+                pass
+
+        try_cfr = hasattr(os, "copy_file_range")
+
+        for offset, count in _get_chunks(srcfd):
+            if count == 0:
+                os.ftruncate(dstfd, offset)
+            else:
+                if try_cfr:
+                    try:
+                        _do_copy_file_range(srcfd, dstfd, offset, count)
```

Comment:
I was wondering why we're doing this / if c_f_r handles the holes part for us, but indeed, the man page says:
> If fd_in is a sparse file, then copy_file_range() may expand any
>       holes existing in the requested range.  Users may benefit from
>       calling copy_file_range() in a loop, and using the [lseek(2)](https://man7.org/linux/man-pages/man2/lseek.2.html)
>       SEEK_DATA and SEEK_HOLE operations to find the locations of data
>       segments.